src/components/TodoInfo/TodoInfo.tsx

```diff
@@ -0,0 +1,229 @@
+import React, {
+  useCallback,
+  useContext,
+  useEffect,
+  useMemo,
+  useRef,
+  useState,
+} from 'react';
+import classNames from 'classnames';
+
+import { deleteTodo, patchTodo } from '../../api/todos';
+import { Todo } from '../../types/Todo';
+import { AppContext } from '../AppContext/AppContext';
+
+type Props = {
+  todo: Todo;
+};
+
+export const TodoInfo: React.FC<Props> = ({ todo }) => {
+  const {
+    id,
+    title,
+    completed,
+  } = todo;
+
+  const [isEditingMode, setIsEditingMode] = useState(false);
+  const [newTitle, setNewTitle] = useState(title);
+
+  const input = useRef<HTMLInputElement>(null);
+
+  const {
+    allTodos,
+    setAllTodos,
+    showError,
+    setShouldShowError,
+    loadingTodosIds,
+    setLoadingTodosIds,
+  } = useContext(AppContext);
+
+  const isLoadingState = useMemo(() => (
+    loadingTodosIds.includes(id)
+  ), [loadingTodosIds]);
+
+  const deleteTodoFromServer = useCallback(async () => {
+    try {
+      setShouldShowError(false);
+      setLoadingTodosIds(prevIds => [...prevIds, id]);
+
+      await deleteTodo(id);
+
+      setAllTodos(prevTodos => prevTodos.filter(prevTodo => (
+        prevTodo.id !== id
+      )));
+    } catch {
+      showError('Unable to delete a todo');
+    } finally {
+      setLoadingTodosIds([0]);
+    }
+  }, [loadingTodosIds, allTodos]);
+
+  const toggleTodoStatus = useCallback(async () => {
+    try {
+      setShouldShowError(false);
+      setLoadingTodosIds(prevIds => [...prevIds, id]);
+
+      await patchTodo(id, { completed: !completed });
+
+      setAllTodos(prevTodos => prevTodos.map(prevTodo => {
+        if (prevTodo.id === id) {
+          return {
+            ...prevTodo,
+            completed: !completed,
+          };
+        }
+
+        return prevTodo;
+      }));
+    } catch {
+      showError('Unable to update a todo');
+    } finally {
+      setLoadingTodosIds([0]);
+    }
+  }, [loadingTodosIds, allTodos]);
+
+  const updateTodoTitle = useCallback(async (updatedTitle: string) => {
+    try {
+      if (updatedTitle !== title) {
+        setShouldShowError(false);
+        setLoadingTodosIds(prevIds => [...prevIds, id]);
+
+        await patchTodo(id, { title: updatedTitle });
+
+        setAllTodos(prevTodos => prevTodos.map(prevTodo => {
+          if (prevTodo.id === id) {
+            return {
+              ...prevTodo,
+              title: updatedTitle,
+            };
+          }
```

Comment:
Use data from server, you can't guarantee that it wasn't changed on server, so better use server data as source of truth